tools/distpartitioning/multi_dev_init.py

```diff
@@ -0,0 +1,499 @@
+import numpy as np
+import math
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+import dgl
+
+from timeit import default_timer as timer
+from datetime import timedelta
+from dataset_utils import get_dataset
+from utils import read_partitions_file, read_json, get_node_types, \
+                    augment_node_data, augment_edge_data, get_ntypes_map, \
+                    write_dgl_objects, write_metadata_json
+from gloo_wrapper import alltoall_cpu_object_lst, alltoallv_cpu, \
+                    alltoall_cpu, allgather_sizes, gather_metadata_json
+from globalids import assign_shuffle_global_nids_nodes, \
+                    assign_shuffle_global_nids_edges, get_shuffle_global_nids_edges
+from convert_partition import create_dgl_object, create_metadata_json, validateDGLObjects
+
+def exchange_node_data(rank, world_size, node_data):
+    """
+    Exchange node_data among the processes in the world
+    Prepare the list of slices targeting each of the process and
+    trigger alltoallv_cpu for the message exchange.
+
+    Parameters:
+    -----------
+    rank : int
+        rank of the current process
+    world_size : int
+        total no. of participating processes
+    node_data : dictionary
+        nodes data dictionary with keys as column names and values as
+        columns from the nodes csv file
+    """
+    input_list = []
+    send_sizes = []
+    recv_sizes = []
+    start = timer()
+    for i in np.arange(world_size):
+        send_idx = (node_data[constants.OWNER_PROCESS] == i)
+        idx = send_idx.reshape(node_data[constants.GLOBAL_NID].shape[0])
+        filt_data = np.column_stack((node_data[constants.NTYPE_ID][idx == 1], \
+                                node_data[constants.GLOBAL_TYPE_NID][idx == 1], \
+                                node_data[constants.GLOBAL_NID][idx == 1]))
+        if(filt_data.shape[0] <= 0): 
+            input_list.append(torch.empty((0,), dtype=torch.int64))
+            send_sizes.append(torch.empty((0,), dtype=torch.int64))
+        else:
+            input_list.append(torch.from_numpy(filt_data))
+            send_sizes.append(torch.tensor(filt_data.shape, dtype=torch.int64))
+        recv_sizes.append(torch.zeros((2,), dtype=torch.int64))
+    end = timer()
+    print('[Rank: ', rank, '] Preparing node_data to send out: ', timedelta(seconds=end - start))
+
+    #exchange sizes first followed by data. 
+    dist.barrier()
+    start = timer()
+    alltoall_cpu(rank, world_size, recv_sizes, send_sizes)
+
+    output_list = []
+    for s in recv_sizes: 
+        output_list.append(torch.zeros(s.tolist(), dtype=torch.int64))
+    
+    dist.barrier()
+    alltoallv_cpu(rank, world_size, output_list, input_list)
+    end = timer()
+    print('[Rank: ', rank, '] Time to exchange node data : ', timedelta(seconds=end - start))
+
+    #stitch together the received data to form a consolidated data-structure
+    rcvd_node_data = torch.cat(output_list).numpy()
```

Comment:
why do you concatenate the data and then assign each column to the dict of node data below?

Reply:
This is after performing alltoallv. the output of this operation is a list of numpy arrays. First I concatenate these individual tensors and convert to numpy arrays to store them in the dictionary, which in this case is node_data. Concatenation is needed because each smaller tensor is received by one of the other ranks in the world.